tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+
+    snapshot_version: Option<NonZeroU64>,
+    snapshot_expires: Option<DateTime<Utc>>,
+    snapshot_extra: Option<HashMap<String, Value>>,
+
+    timestamp_version: Option<NonZeroU64>,
+    timestamp_expires: Option<DateTime<Utc>>,
+    timestamp_extra: Option<HashMap<String, Value>>,
+}
+
+impl RepositoryEditor {
+    /// Create a new, bare `RepositoryEditor`
+    pub fn new<P>(root_path: P) -> Result<RepositoryEditor>
+    where
+        P: AsRef<Path>,
+    {
+        // Read and parse the root.json. Without a good root, it doesn't
+        // make sense to continue
+        let root_path = root_path.as_ref();
+        let root_buf = std::fs::read(root_path).context(error::FileRead { path: root_path })?;
+        let root_buf_len = root_buf.len() as u64;
+        let root = serde_json::from_slice::<Signed<Root>>(&root_buf)
+            .context(error::FileParseJson { path: root_path })?;
+        let mut digest = [0; SHA256_OUTPUT_LEN];
+        digest.copy_from_slice(ring::digest::digest(&SHA256, &root_buf).as_ref());
+
+        let signed_root = SignedRole {
+            signed: root,
+            buffer: root_buf,
+            sha256: digest,
+            length: root_buf_len,
+        };
+
+        Ok(RepositoryEditor {
+            signed_root,
+            new_targets: None,
+            existing_targets: None,
+            targets_version: None,
+            targets_expires: None,
+            targets_extra: None,
+            snapshot_version: None,
+            snapshot_expires: None,
+            snapshot_extra: None,
+            timestamp_version: None,
+            timestamp_expires: None,
+            timestamp_extra: None,
+        })
+    }
+
+    /// Given a `tough::Repository` and the path to a valid root.json, create a
+    /// `RepositoryEditor`. This `RepositoryEditor` will include all of the targets
+    /// and bits of _extra metadata from the roles included. It will not, however,
+    /// include the versions or expirations and the user is expected to set them.
+    pub fn from_repo<T, P>(root_path: P, repo: Repository<'_, T>) -> Result<RepositoryEditor>
+    where
+        P: AsRef<Path>,
+        T: Transport,
+    {
+        let mut editor = RepositoryEditor::new(root_path)?;
+        editor.targets(repo.targets.signed)?;
+        editor.snapshot(repo.snapshot.signed)?;
+        editor.timestamp(repo.timestamp.signed)?;
+
+        Ok(editor)
+    }
+
+    /// Add an existing `Targets` struct to the repository. Any existing
+    /// targets are preserved.
+    pub fn targets(&mut self, targets: Targets) -> Result<&mut Self> {
+        ensure!(
+            targets.spec_version == SPEC_VERSION,
+            error::SpecVersion {
+                given: targets.spec_version,
+                supported: SPEC_VERSION
+            }
+        );
+
+        // Hold on to the existing targets
+        self.existing_targets = Some(targets.targets);
+        self.targets_extra = Some(targets._extra);
```

Comment:
Do we need methods to set *_extra fields?  Currently these methods that fully replace exiting targets/bla are the only way to set it.

Reply:
I'm not opposed to it, but I'm wondering how much of a use case it will really have.

Reply:
I agree, I'd rather leave it out.  I hope no one would rely on those extra fields just because they can set them when deserializing.  (I hope we can eliminate them at some point...)